plugin/vscode/performance-writer.js

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2020  Matthew Glazar
+// See end of file for extended copyright information.
+
+const fs = require('fs');
+const path = require("path");
+const process = require("process");
+
+class PerformanceWriter {
+    
+    constructor (extensionContext) {
+        this._performanceTxtPath = path.join(process.cwd(), "test_performance.txt");
+        this._dataAttributes = {
+            columns:'timestamp,file-ref,file-name-or-duration,event-id'
+        };
+        this._fileName;
+        this._fileIndex;
+        this._extensionContext = extensionContext;
+    }
+
+    fileNameOptimizer (){
+        let currentSessionMap = this._extensionContext.globalState.get('qljsFileMap');
+        this._fileName = this._fileName.split(path.sep);
+        this._fileName = this._fileName[this._fileName.length - 1];
+        if (!(currentSessionMap)) {
+            currentSessionMap = {};
+            currentSessionMap['latestIndex'] = 1;
+            currentSessionMap[this._fileName] = 1;
+            this._fileIndex = 1;
+            this.writeToFile(null,"header");
+            this.writeToFile(null, "newFile");
+            this._extensionContext.globalState.update('qljsFileMap', currentSessionMap);
+        } else {
+            let existingFileIndex = currentSessionMap[this._fileName];
+            if (existingFileIndex) {
+                this._fileIndex = existingFileIndex;
+            } else {
+                currentSessionMap['latestIndex'] = currentSessionMap['latestIndex'] + 1;
+                currentSessionMap[this._fileName] = currentSessionMap['latestIndex'];
+                this._fileIndex = currentSessionMap['latestIndex'];
+                this._extensionContext.globalState.update('qljsFileMap', currentSessionMap);
+                this.writeToFile(null, "newFile", currentSessionMap['latestIndex']);
+            }
+        }
+    }
+    
+    writeToFile(timeTaken=null, mode=null) {
+        let tracing_line;
+
+        if (mode === 'header') {
+            tracing_line = this._dataAttributes.columns;
+        } else if (mode === "newFile") {
+            tracing_line = '\n' + [Date.now(), this._fileIndex, this._fileName, 'n'].join(',');
+        } else {
+            tracing_line = '\n' + [Date.now(), this._fileIndex, timeTaken, ''].join(',');
+        }
+
+        fs.appendFile(this._performanceTxtPath, tracing_line, (err) => {
```

Comment:
Blocking: This is problematic for two reasons:
1. I assume that I/O is very expensive compared to linting. Here, we're opening a file, writing a tiny piece of data, then closing the file, every time the user types something. It shouldn't be blocking, so it shouldn't hang the editor or preventing linting from happening, but it's still wasteful of energy.
2. This is asynchronous. Are appends guaranteed to happen in order? Are appends guaranteed not to interleave? If not, we'll corrupt the file.

I suggest we do the following instead:
1. When we need to log data, append records to an array.
2. Periodically (e.g. every second), if there are any records in the array, append all the records to the file.

Reply:
did it according to your suggested way in latest commit